### PR TITLE
feat: quick actions

### PR DESCRIPTION
**Pull request by voicis**
Enable Quick Action List in CPE for Adaptation editor.
LR/ALP
Enable Clear button in the filter bar
Extend a page with an own controller
Change table columns

Object Page:
Change table columns
Extend a page with an own controller

### CODE REVIEW

File: packages/preview-middleware-client/src/cpe/quick-actions/definitions/lr-enable-clear-filter-bar.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+            const command = await CommandFactory.getCommandFor<FlexCommand>(
+                modifiedControl,
+                'Property',
+                modifiedValue,
+                null,
+                flexSettings
+            );
+
+            await context.rta.getCommandStack().pushAndExecute(command);
```
Comment (Jimmy-Joseph19) [inline packages/preview-middleware-client/src/cpe/quick-actions/definitions/lr-enable-clear-filter-bar.ts:50]: Return the command and execute in QuickActionService. So removes rta

File: packages/preview-middleware-client/src/adp/quick-actions/common/add-controller-to-page.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
+import OverlayRegistry from 'sap/ui/dt/OverlayRegistry';
+import FlexCommand from 'sap/ui/rta/command/FlexCommand';
+import UI5Element from 'sap/ui/core/Element';
+
+import { SIMPLE_QUICK_ACTION_KIND, SimpleQuickAction } from '@sap-ux-private/control-property-editor-common';
+
+import { getUi5Version } from '../../../utils/version';
+import { getAllSyncViewsIds, getControllerInfoForControl } from '../../utils';
+import { getRelevantControlFromActivePage } from '../../../cpe/quick-actions/utils';
```
Comment (heimwege) [inline packages/preview-middleware-client/src/adp/quick-actions/common/add-controller-to-page.ts:9]: If `getRelevantControlFromActivePage` is needed in cpe and adp shouldn't this then be moved to `src/utils` or somewhere more central at src?

File: packages/preview-middleware-client/src/adp/quick-actions/common/op-add-header-field.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+import UI5Element from 'sap/ui/core/Element';
+import ObjectPageLayout from 'sap/uxap/ObjectPageLayout';
+import FlexBox from 'sap/m/FlexBox';
+
+import { SIMPLE_QUICK_ACTION_KIND, SimpleQuickAction } from '@sap-ux-private/control-property-editor-common';
+
+import { DialogNames, handler } from '../../../adp/init-dialogs';
+
+import { getRelevantControlFromActivePage } from '../../../cpe/quick-actions/utils';
+import { QuickActionContext, SimpleQuickActionDefinition } from '../../../cpe/quick-actions/quick-action-definition';
```
Comment (heimwege) [inline packages/preview-middleware-client/src/adp/quick-actions/common/op-add-header-field.ts:12]: If quick actions are needed in cpe and adp shouldn't this then be moved to src/utils or somewhere more central at src?

**Review by heimwege — COMMENTED** [review]
See Comments. Additionally there is almost no JSDoc for most of the new functions (e.g. quick-actions/**). Is this intended?

File: packages/preview-middleware-client/src/adp/quick-actions/common/op-add-header-field.ts (hunk 3 above)
Comment (voicis) [reply packages/preview-middleware-client/src/adp/quick-actions/common/op-add-header-field.ts:12]: The current structure is as follows:
- `flp` code that is required for all previews
- `cpe` RTA plugin for working with CPE
- `adp` extends `cpe` plugin with adaptation specific code

That means `src/utils` probably would need to contain very generic code that would also make sense in scope of `flp`. For things that make sense only in RTA and CPE scope we so far keep in the `cpe` folder.

We probably should move the reuse code a bit higher and make it more clear which part of `cpe` is intended for reuse/extension by `adp`.

Comment (heimwege) [reply packages/preview-middleware-client/src/adp/quick-actions/common/op-add-header-field.ts:12]: Thanks for the explanation. This restructuring sounds like a task that does not necessarily need to be done in this PR.

File: packages/preview-middleware-client/test/unit/cpe/init.test.ts [hunk 4]
```diff
@@ -64,28 +79,31 @@ describe('main', () => {
...
+    //     getSelection: jest.fn().mockReturnValue([{ setSelected: jest.fn() }, { setSelected: jest.fn() }]),
+    //     attachUndoRedoStackModified: jest.fn(),
+    // getFlexSettings: jest.fn().mockReturnValue({ layer: 'VENDOR', scenario: common.SCENARIO.UiAdaptation }),
+    //     getRootControlInstance: jest.fn().mockReturnValue({
+    //         getManifest: jest.fn().mockReturnValue({ 'sap.app': { id: 'testId' } })
+    //     }),
+    //     attachStop: jest.fn(),
+    //     getService: jest.fn(),
+    //     attachModeChanged: jest.fn()
+    // } as any;
```
Comment (nikmace) [inline packages/preview-middleware-client/test/unit/cpe/init.test.ts:97]: Will the comments be remove or this is needed in the future?

**Review by nikmace — CHANGES_REQUESTED** [review]
Overall it looks good and a lot of effort was put into this PR which is much appreciated. 
I only have some cosmetic comments otherwise.

There are also 10 new issues to solve.

Also, we would like to get this PR tested after the reviews pass (before it gets merged).

File: packages/preview-middleware-client/src/cpe/quick-actions/utils.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+ * @param page - Page control.
+ * @param controlId - UI5 control id.
+ * @returns True if control is visible in the page.
+ */
+export function pageHasControlId(page: Control, controlId: string): boolean {
+    const controlDomElement = getControlById(controlId)?.getDomRef();
+    return !!controlDomElement && !!page?.getDomRef()?.contains(controlDomElement);
+}
+
+function isDescendantOfPage(control: ManagedObject | null | undefined, oRootControl: ManagedObject) {
```
[the remark this answers is not in the document]
Comment (Jimmy-Joseph19) [reply packages/preview-middleware-client/src/cpe/quick-actions/utils.ts:31]: added

File: packages/preview-middleware-client/src/cpe/quick-actions/quick-action-service.ts [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+                    }
+                }
+                groups.push(group);
+            }
+        }
+
+        this.sendAction(quickActionListChanged(groups));
+    }
+
+    private addAction(group: QuickActionGroup, instance: QuickActionDefinition): void {
```
[the remark this answers is not in the document]
Comment (Jimmy-Joseph19) [reply packages/preview-middleware-client/src/cpe/quick-actions/quick-action-service.ts:124]: Usually we add jsdocs for public/export methods

File: packages/control-property-editor/src/panels/RightPanel.scss [hunk 7]
```diff
@@ -0,0 +1,12 @@
+.editor__properties {
+    .section__body {
+        display: flex;
+        flex-direction: column;
+        margin-bottom: 0px;
+        margin-top: 0px;
+        margin-left: 0px;
+        margin-right: 0px;
```
[the remark this answers is not in the document]
Comment (Jimmy-Joseph19) [reply packages/control-property-editor/src/panels/RightPanel.scss]: adapted

File: packages/control-property-editor/src/panels/RightPanel.scss [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+        padding-left: 0px !important;
```
[the remark this answers is not in the document]
Comment (Jimmy-Joseph19) [reply packages/control-property-editor/src/panels/RightPanel.scss:6]: yes

File: packages/preview-middleware-client/src/adp/quick-actions/common/op-add-header-field.ts [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+import { isA } from '../../../utils/core';
+export const OP_ADD_HEADER_FIELD_TYPE = 'op-add-header-field';
+const CONTROL_TYPES = ['sap.uxap.ObjectPageLayout'];
+
+/**
+ * Quick Action for adding a Header Field to an Object Page.
+ */
+export class AddHeaderFieldQuickAction implements SimpleQuickActionDefinition {
```
[the remark this answers is not in the document]
Comment (voicis) [reply packages/preview-middleware-client/src/adp/quick-actions/common/op-add-header-field.ts:20]: Interfaces should have the documentation for the public methods, but it looks like it currently is missing and needs to be added.

File: packages/preview-middleware-client/test/unit/cpe/init.test.ts (hunk 4 above)
Comment (voicis) [reply packages/preview-middleware-client/test/unit/cpe/init.test.ts]: Removed 968665e

**Review by marufrasully — DISMISSED** [review]
Reviewed following packages
* @sap-ux-private/control-property-editor-common
* @sap-ux/control-property-editor
Changes looks good. Did not test it locally.

**Review by nikmace — COMMENTED** [review]
Please add the following
- description of the PR (changes)
- the link to the issue

Otherwise, the code looks good. Still need to test this locally.

**Review by nikmace — APPROVED** [review]
Code looks good
Coverage is OK
Previewed locally but **did not test**

**Review by marufrasully — APPROVED** [review]
Reviewed following packages
* @sap-ux-private/control-property-editor-common
* @sap-ux/control-property-editor
Changes looks good. Did not test it locally.

**Pull request merged**